### PR TITLE
feat(device-discovery): module discovery for nokia_sros / nokia_sros_ssh (batch 5)

### PR DESCRIPTION
**Pull request by leoparente**
## Summary

- Adds `get_modules()` to `nokia_sros` (NETCONF) and `nokia_sros_ssh` (SSH CLI) custom NAPALM drivers.
- Emits the chassis → card → MDA → transceiver tree (depth 3) for modular Nokia 7750 SR-7/SR-12/SR-12e, 7750 SR-7s/SR-14s, and 7950 XRS-20 chassis through the canonical `_modules.py` envelope.
- Both transports produce a byte-identical envelope for the same physical chassis (parity asserted in tests across sr12_full, sr7s_imm, sr1_fixed scenarios).

**Card classification (Q1=B from spec brainstorm):** IOM/IMM → `linecard`; CPM → `supervisor`; SFM → `linecard` (matches batch-2 EOS / batch-3 NX-OS fabric handling). Fixed-form-factor SR-1 / SR-1s emit no module entries.

**Approach:** Per-driver bespoke (Approach A) — no shared `_nokia_common.py`. NETCONF uses two ncclient subtree filters against `urn:nokia.com:sros:ns:yang:sr:state`. SSH uses driver-local regex over `show chassis|card|mda|port detail` (no ntc-templates exist for these commands). Single-member envelope keyed by `None` — SR-OS chassis are not stacked.

## Review history

This branch went through four rounds of adversarial review. Each round dispatched a fresh claude subagent over the full branch diff; findings were triaged and fixed before opening this PR.

| Round | Findings | Status |
|------|----------|--------|
| 1 | 3 BLOCKER, 6 IMPORTANT, 5 NIT | All fixed |
| 2 | 1 BLOCKER, 2 IMPORTANT, 3 NIT | All fixed; 3 NIT documented |
| 3 | 0 BLOCKER, 1 IMPORTANT, 2 NIT | All fixed |
| 4 | 0 BLOCKER, 0 IMPORTANT, 5 NIT | 2 NIT fixed; 3 NIT out of scope |

Notable defects shaken out by review:
- MD-CLI / SR-OS 19R8+ field-label case variance (e.g. `Card type` vs `Card Type`)
- Provisioned-but-not-equipped slots silently dropped on consecutive headers
- SR-OS `MINOR:` / `Error:` text being fed to the transceiver parser
- NETCONF masking `AttributeError` as transport failure
- Full-mode interface routing missing per-transceiver `interfaces_by_bay` key
- Brute-force ~1100-command port scan replaced with single `show port` enumeration
- IMM cards initially treated as leaf — corrected to expose integrated MDA
- FP4 connector-cage (`1/1/c2/1`) and QSFP breakout (`1/1/1[1]`) port-id forms
- AAA-blocked `show port` silently emptying the transceiver pass

## Test plan

- [x] `pytest tests/` — 1558 passing, 134 skipped (+22 vs develop baseline)
- [x] `ruff check .` — clean
- [x] Parity assertions across all three scenarios (sr12_full, sr7s_imm, sr1_fixed)
- [x] Unit tests for SSH port-list regex (classic, c-cage, breakout, last-line, line-internal rejection)
- [x] Unit tests for NETCONF transceiver attachment (classic, c-cage, unknown-MDA)
- [x] No `OBS-####`/`ENGHLP-####`/`NetBox module` strings in diff
- [ ] Real device validation (no mockit ssh template for `show card detail` / `show mda detail` yet — covered by realistic fixtures based on Nokia docs)

orb-agent docs update will land via PR #381 (existing draft tracking batch 3+4 docs).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/nokia_sros.py [hunk 1]
```diff
@@ -387,6 +452,247 @@ def _parse_last_flapped(flap_str: str) -> float:
...
+        if mda_bay is None or mda_bay.module is None:
+            continue
+        mda_bay.module.sub_bays.append(_ModuleBay(
+            name=port_id, position=port_id,
+            module=_ModuleEntry(model=model, serial=sn, type="transceiver", description=""),
+        ))
+        # Emit BOTH the MDA-path key (linecards mode → interface routes to
+        # MDA module) and the per-port key (full mode → deepest-bay-wins
+        # routes interface to its specific transceiver sub-bay).
+        interfaces_by_bay.setdefault(mda_path, []).append(port_id)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros.py:677]: Valid finding — fixed in a88e14b. Both drivers now emit a card-slot key
in `interfaces_by_bay` aggregating all ports under that card, in
addition to the MDA-path and per-port keys. The translator's
deepest-bay-wins resolution still attaches the interface to its specific
transceiver in `full` mode (per-port key overwrites MDA-path which
overwrites card-slot), while `linecards` mode now correctly routes the
interface to the parent linecard module via the card-slot key it
consults at depth 1.

Added a unit test (`test_attach_transceiver_card_slot_key_aggregates_across_mdas`)
verifying the aggregation across MDAs on the same card.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/nokia_sros.py [hunk 2]
```diff
@@ -387,6 +452,252 @@ def _parse_last_flapped(flap_str: str) -> float:
...
+    """
+    Classify a Nokia SR-OS card / MDA by its equipped-type string.
+
+      iom*, imm*  -> linecard
+      cpm*        -> supervisor
+      sfm*        -> linecard (no separate fabric type today)
+      anything else -> "other" (envelope drops on emit)
+    """
+    et = (equipped_type or "").strip().lower()
+    if et.startswith(("iom", "imm")):
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros.py]: Valid finding — fixed in 33adcf0. Both drivers' classifiers now accept `xcm*` as `linecard`, so 7950 XRS chassis emit their XCM forwarding cards (with XMAs/C-XMAs picked up automatically via the existing MDA path). Added test_classify_xcm_as_linecard and a regression test covering the full prefix table.

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 3]
```diff
@@ -144,6 +157,355 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+
+_NOKIA_SROS_SEP_RE = re.compile(r"^=+\s*$", re.MULTILINE)
+_NOKIA_SROS_FIELD_RE = re.compile(
+    r"^\s*(?P<label>[A-Za-z][\w \-/.()]*?)\s*:\s*(?P<value>.+?)\s*$",
+    re.MULTILINE,
+)
+# "Card 1 Detail" or "Card A Detail".
+_NOKIA_SROS_CARD_HDR_RE = re.compile(r"Card\s+(?P<slot>[A-Za-z0-9]+)\s+Detail")
+# "MDA 1/1 Detail".
+_NOKIA_SROS_MDA_HDR_RE = re.compile(r"MDA\s+(?P<slot>\d+)/(?P<mda>\d+)\s+Detail")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py]: Valid finding — fixed in 33adcf0. All three SSH detail-header regexes (Card / MDA / Port) now use `re.IGNORECASE`, so MD-CLI `MDA 4/1 detail` matches the same as classic-CLI `MDA 4/1 Detail`. Added test_parse_cards_case_insensitive_detail_header and test_parse_mdas_case_insensitive_detail_header to lock this in.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 4]
```diff
@@ -144,6 +157,362 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+    pending_port: str | None = None
+    for block in _nokia_sros_ssh_split_blocks(text):
+        hdr = _NOKIA_SROS_PORT_HDR_RE.search(block)
+        if hdr is not None:
+            pending_port = hdr.group("port")
+            continue
+        if pending_port is None:
+            continue
+        fields = _nokia_sros_ssh_extract_fields(block)
+        model = fields.get("model", "")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py]: Valid finding — fixed in 95e40ba. The transceiver parser now accepts both `Model Number` (real SR-OS, per the Nokia command reference you linked) and the bare `Model` label as a fallback. The integration fixtures were also updated to use the real-world `Model Number :` label so parity tests now exercise the authentic format. Two new unit tests (`test_parse_port_transceiver_accepts_model_number_label`, `test_parse_port_transceiver_accepts_bare_model_label`) lock both forms in.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 5]
```diff
@@ -144,6 +157,365 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+            pending_slot = hdr.group("slot")
+            continue
+        if pending_slot is None:
+            continue
+        fields = _nokia_sros_ssh_extract_fields(block)
+        rows.append({
+            "slot": pending_slot,
+            "equipped_type": fields.get("card type", ""),
+            "pid": fields.get("part number", ""),
+            "sn": fields.get("serial number", ""),
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:309]: Valid finding — fixed in eb4c423. The card parser now treats the Card Summary table as the source of truth for `equipped_type` and the per-card Hardware Data subsection as the source of `Part number` / `Serial number`. Header regex relaxed to accept `Card N` (no `Detail` keyword), with slot pattern tightened to `[A-Za-z]\d*|\d+` so `Card Summary` is not mis-matched as slot=`Summary`. Fixtures realigned to the realistic classic-CLI format. Two new unit tests (`test_parse_cards_real_sros_format_uses_summary_for_equipped_type`, `test_parse_cards_summary_header_not_mismatched_as_slot`).

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 6]
```diff
@@ -144,6 +157,365 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+        driver.device.send_command("show chassis detail")
+    except Exception as e:
+        logger.warning("nokia_sros_ssh.get_modules: show chassis failed: %s", e)
+    try:
+        card_raw = driver.device.send_command("show card detail")
+    except Exception as e:
+        logger.warning("nokia_sros_ssh.get_modules: show card failed: %s", e)
+        return None
+    try:
+        mda_raw = driver.device.send_command("show mda detail")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py]: Valid finding — fixed in eb4c423. The driver no longer issues bare `show mda detail` (invalid SR-OS syntax). New helper `_nokia_sros_ssh_fetch_and_parse_mdas` iterates each linecard-classified card from `show card detail` and issues `show mda <slot> detail` per slot, concatenating the parsed rows. CPM and any other non-linecard slots are skipped. Fixtures realigned to per-slot `show_mda_<N>_detail.txt` files for SR-12 and SR-7s.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 7]
```diff
@@ -144,6 +157,434 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+_NOKIA_SROS_MDA_HDR_RE = re.compile(
+    r"^MDA\s+(?P<slot>\d+)/(?P<mda>\d+)(?:\s+Detail)?\s*$",
+    re.IGNORECASE | re.MULTILINE,
+)
+# `show card` summary-table rows: each row has slot, provisioned-type,
+# equipped-type, admin-state, oper-state. Used as the source of truth for
+# equipped_type since SR-OS per-card detail blocks may not repeat that
+# field (they expose Part/Serial inside a Hardware Data subsection).
+_NOKIA_SROS_CARD_SUMMARY_RE = re.compile(
+    r"^(?P<slot>[A-Za-z]\d*|\d+)\s+(?P<prov>\S+)\s+(?P<equipped>\S+)\s+(?:up|down)\s+(?:up|down)",
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py]: Valid finding — fixed in ff1395d. The summary regex's `equipped` capture is now optional (with a negative lookahead so `up`/`down` cannot be mis-captured), and state tokens accept an optional `/<suffix>` to handle CPM rows like `up/active` and `up/standby`. When the equipped column is blank (the documented Nokia behavior when it matches the provisioned type), the parser falls back to the provisioned-type token. Unit test `test_parse_card_summary_single_type_row` covers both row forms and the state-suffix shapes.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/nokia_sros.py [hunk 8]
```diff
@@ -387,6 +452,252 @@ def _parse_last_flapped(flap_str: str) -> float:
...
+        model_el = tx.find("state_ns:model", _NSMAP)
+        sn_el = tx.find("state_ns:serial-number", _NSMAP)
+        part_el = tx.find("state_ns:part-number", _NSMAP)
+        port_id = (port_id_el.text or "").strip()
+        model = (model_el.text or "").strip() if model_el is not None else ""
+        sn = (sn_el.text or "").strip() if sn_el is not None else ""
+        if not model and part_el is not None:
+            model = (part_el.text or "").strip()
+        if not (port_id and model and sn):
+            continue
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros.py]: Valid finding — fixed in ba9c627. Separated routing emission from transceiver sub-bay emission in both transports. Every discovered port under a known MDA now gets card-slot and mda-path routing entries in `interfaces_by_bay`; the transceiver sub-bay (and its per-port deepest-wins key) is emitted only when both model and serial are populated. Fixtures realigned to include optic-less ports (1/1/2, 2/1/1) so the parity tests now exercise the copper / empty-cage code path. New unit test `test_attach_transceiver_routes_optic_less_port_to_parent_bays` locks the separation in.

**Comment by leoparente** [comment]
@codex review

File: device-discovery/custom_napalm/nokia_sros.py [hunk 9]
```diff
@@ -180,6 +194,57 @@
...
+            </hardware-data>
+            <mda>
+                <mda-slot/>
+                <equipped-type/>
+                <hardware-data>
+                    <part-number/>
+                    <serial-number/>
+                </hardware-data>
+            </mda>
+        </card>
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros.py:225]: Valid finding — fixed in 40fbdc7. The NETCONF `_FILTER_MODULES` now includes the `<sfm>` subtree, and the row walker emits one top-level bay per SFM with the slot prefixed `SFM <N>` to avoid collision with same-numbered card slots. The existing classifier (`sfm*` → linecard) handles them downstream. SR-12 fixture extended with three SFMs (`sfm5-12`); expected envelope now lists `SFM 1`, `SFM 2`, `SFM 3` after the IOM bays. New unit test `test_rows_from_state_xml_includes_sfm_subtree`.

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 10]
```diff
@@ -144,6 +157,452 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+    try:
+        driver.device.send_command("show chassis detail")
+    except Exception as e:
+        logger.warning("nokia_sros_ssh.get_modules: show chassis failed: %s", e)
+    try:
+        card_raw = driver.device.send_command("show card detail")
+    except Exception as e:
+        logger.warning("nokia_sros_ssh.get_modules: show card failed: %s", e)
+        return None
+    card_rows = _nokia_sros_ssh_parse_cards(card_raw or "")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:652]: Valid finding — fixed in 40fbdc7. The SSH driver now issues `show sfm detail` after `show card detail` and parses the output with a dedicated `_nokia_sros_ssh_parse_sfms` helper. Parsed SFM rows are appended to `card_rows` (with `SFM <N>` slot names) before the MDA fetch, so the existing assembler treats them as top-level bays via the established `sfm*` → linecard classifier path. New `show_sfm_detail.txt` fixture for SR-12 and unit test `test_parse_sfms_real_sros_format`.

**Comment by leoparente** [comment]
@codex review

**Review by samiura — APPROVED** [review]

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 11]
```diff
@@ -144,6 +157,444 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+
+def _nokia_sros_ssh_get_modules_impl(driver) -> dict | None:
+    """
+    Module discovery for Nokia SR-OS via SSH CLI.
+
+    Issues `show chassis|card|mda detail` then iterates each parsed MDA's
+    port range, issuing `show port slot/mda/N detail` to pick up optics.
+    FakeCLIDevice returns "" for missing files so the scan is cheap in
+    tests; real hardware enumerates only valid port ids.
+    """
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:639]: Valid finding — fixed in c66156f. Docstring rewritten to reflect the actual flow: `show port` enumerates port ids, then `show port <port-id> detail` is issued only for ports whose slot/mda matches a known MDA. The full command sequence + bound is now documented inline.

File: device-discovery/custom_napalm/nokia_sros.py [hunk 12]
```diff
@@ -27,6 +27,20 @@
 from lxml import etree
 from napalm.base import models
 from napalm.base.helpers import convert
+from ncclient import NCClientError
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros.py:31]: Valid — fixed in c66156f. File-level docstring now lists `get_modules` alongside the other implemented getters.

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 13]
```diff
@@ -19,6 +19,19 @@
 from napalm.base.netmiko_helpers import netmiko_args
 from ntc_templates.parse import parse_output
 
+from custom_napalm._modules import (
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:23]: Valid — fixed in c66156f. File-level docstring now lists `get_modules` alongside the other implemented getters.

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 14]
```diff
@@ -144,6 +157,452 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+
+def _nokia_sros_ssh_get_modules_impl(driver) -> dict | None:
+    """
+    Module discovery for Nokia SR-OS via SSH CLI.
+
+    Issues `show chassis|card|mda detail` then iterates each parsed MDA's
+    port range, issuing `show port slot/mda/N detail` to pick up optics.
+    FakeCLIDevice returns "" for missing files so the scan is cheap in
+    tests; real hardware enumerates only valid port ids.
+    """
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:639]: Valid — fixed in c66156f. Docstring now reflects the show-port-enumeration flow and the bounded command count.

File: device-discovery/custom_napalm/nokia_sros.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros.py:31]: Valid — fixed in c66156f. File-level docstring now lists `get_modules`.

File: device-discovery/custom_napalm/nokia_sros_ssh.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:23]: Valid — fixed in c66156f. File-level docstring now lists `get_modules`.

File: device-discovery/custom_napalm/nokia_sros_ssh.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:639]: Valid — fixed in c66156f. Docstring rewritten to reflect the show-port-enumeration flow.

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 15]
```diff
@@ -144,6 +157,503 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+    re.IGNORECASE | re.MULTILINE,
+)
+_NOKIA_SROS_MDA_HDR_RE = re.compile(
+    r"^MDA\s+(?P<slot>\d+)/(?P<mda>\d+)(?:\s+Detail)?\s*$",
+    re.IGNORECASE | re.MULTILINE,
+)
+# `SFM 1`, `SFM 1 detail`, etc. SFMs live in a separate Nokia state path
+# from cards; SSH queries them via `show sfm`.
+_NOKIA_SROS_SFM_HDR_RE = re.compile(
+    r"^SFM\s+(?P<slot>\d+)(?:\s+Detail)?\s*$",
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py]: Valid finding — fixed in c66156f. The SFM-header regex now accepts both `Fabric <N>` (per the Nokia command reference for `show sfm <id> detail`) and the legacy `SFM <N>` form, case-insensitively. SR-12 SSH fixture switched to the realistic `Fabric N` shape; unit test `test_parse_sfms_real_sros_format` now exercises both header variants in one parse call.

File: device-discovery/custom_napalm/nokia_sros_ssh.py [hunk 16]
```diff
@@ -144,6 +157,503 @@ def _parse_port_hw_mac_addresses(text: str) -> dict[str, str]:
...
+    all_rows: list[dict] = []
+    for row in card_rows:
+        if classify_module_type_nokia_sros_ssh(row.get("equipped_type", "")) != "linecard":
+            continue
+        slot = row.get("slot") or ""
+        if not slot:
+            continue
+        try:
+            raw = driver.device.send_command(f"show mda {slot} detail")
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (leoparente) [reply device-discovery/custom_napalm/nokia_sros_ssh.py:553]: Valid — fixed in 2787d21. The MDA-fetch loop now keys on the equipped_type prefix (iom* / imm* / xcm*) rather than the broader linecard classification, so SFM and CPM rows are skipped at the source — no `show mda SFM <N> detail` round-trips. New unit test `test_fetch_mdas_skips_sfm_and_cpm_slots` locks the filter in.

**Comment by leoparente** [comment]
@codex review
